pkg/k8smgmt/appinst.go

```diff
@@ -60,86 +95,54 @@ func LbServicePortToString(p *v1.ServicePort) string {
+	cmd := fmt.Sprintf("kubectl %s get pods -n %s --selector=%s -o json", kConfArg, namespace, selector)
 	out, err := client.Output(cmd)
 	if err != nil {
 		log.SpanLog(ctx, log.DebugLevelInfra, "error getting pods", "err", err, "out", out)
 		return done, fmt.Errorf("error getting pods: %v", err)
 	}
-	lines := strings.Split(out, "\n")
-	// there are potentially multiple pods in the lines loop, we will quit processing this obj
-	// only when they are all up, i.e. no non-
-	podCount := 0
-	runningCount := 0
-
-	for _, line := range lines {
-		if line == "" {
-			continue
-		}
-		if strings.Contains(line, "No resources found") {
-			// If creating, pods may not have taken
-			// effect yet. If deleting, may already
-			// be removed.
-			if waitFor == WaitRunning && time.Since(startTimer) > createWaitNoResources {
-				return done, fmt.Errorf("no resources found for %s on create: %s", createWaitNoResources, line)
-			}
-			break
-		} else if podStateReg.MatchString(line) {
-			// there can be multiple pods, one per line. If all
-			// of them are running we can quit the loop
-			podCount++
-			matches := podStateReg.FindStringSubmatch(line)
-			podName := matches[1]
-			podState := matches[2]
-			reason := matches[3]
-			message := matches[4]
-			switch podState {
-			case "Running":
-				log.SpanLog(ctx, log.DebugLevelInfra, "pod is running", "podName", podName)
-				runningCount++
-			case "Pending":
-				if reason == "Unschedulable" {
-					log.SpanLog(ctx, log.DebugLevelInfra, "pod cannot be scheduled", "podName", podName, "message", message)
-					return done, fmt.Errorf("Run container failed, pod could not be scheduled, message: %s", message)
-				}
-				fallthrough
-			case "ContainerCreating":
-				fallthrough
-			case "CreateContainerConfigError": // this can be a transient state for some deployments
-				log.SpanLog(ctx, log.DebugLevelInfra, "still waiting for pod", "podName", podName, "state", podState)
-			case "Terminating":
-				log.SpanLog(ctx, log.DebugLevelInfra, "pod is terminating", "podName", podName, "state", podState)
-			default:
-				log.SpanLog(ctx, log.DebugLevelInfra, "pod state unhandled", "podName", podName, "state", podState, "out", out)
-				if podState == "Failed" && waitFor == WaitDeleted {
-					// Failed state can happen momentarily when
-					// pod's container is in state Terminated,
-					// before it's actually removed. If we hit this
-					// while waiting for a delete, let it try again
-					continue
-				}
-				if strings.Contains(podState, "Init") {
-					// Init state cannot be matched exactly, e.g. Init:0/2
-					log.SpanLog(ctx, log.DebugLevelInfra, "pod in init state", "podName", podName, "state", podState)
+	podList := v1.PodList{}
+	if err := json.Unmarshal([]byte(out), &podList); err != nil {
```

Comment:
nice - finally structured data not string-parsing :)

Reply:
I know, this thing was super flakey. This code is btw based off of the same function in our edgecloud operator, so it's pretty well tested already.